flocker/node/agents/blockdevice.py

```diff
@@ -1639,311 +1647,95 @@ def _calculate_desired_for_manifestation(self, manifestation):
-                if manifestation.dataset.deleted is False:
-                    # It is supposed to be here.
-                    continue
-                else:
-                    # Detatch it so it can be deleted.
-                    pass
-            if unicode(attached_dataset_id) in paths:
-                # It is mounted and needs to unmounted before it can be
-                # detached.
+            if dataset.state != DatasetStates.MOUNTED:
+                # A lease doesn't force a mount.
                 continue
-            yield DetachVolume.from_state_and_config(
-                discovered_dataset=discovered_datasets[attached_dataset_id],
-                desired_dataset=DesiredDataset(dataset_id=attached_dataset_id),
+            # This may override something from above, if there is a
+            # lease or application using a dataset.
+            desired_datasets[dataset_id] = DesiredDataset(
+                dataset_id=dataset_id,
+                state=DatasetStates.MOUNTED,
+                maximum_size=dataset.maximum_size,
+                # XXX We don't populate metadata here, but it isn't necessary
+                # until we want to update it.
+                metadata={},
+                mount_point=self._mountpath_for_dataset_id(
+                    unicode(dataset_id)
+                ),
             )
 
-    def _calculate_attaches(self, devices, configured, nonmanifest, volumes,
-                            discovered_datasets):
-        """
-        :param PMap devices: The datasets with volumes attached to this node
-            and the device files at which they are available.  This is the same
-            as ``NodeState.devices``.
-        :param PMap configured: The manifestations which are configured on this
-            node.  This is the same as ``NodeState.manifestations``.
-        :param PMap nonmanifest: The datasets which exist in the cluster but
-            are not attached to any node.
-        :param volumes: An iterable of ``BlockDeviceVolume`` instances that are
-            known to exist in the cluster.
-
-        :return: A generator of ``AttachVolume`` instances, one for each
-                 dataset which exists, is unattached, and is configured to be
-                 attached to this node.
-        """
-        for manifestation in configured.values():
-            dataset_id = UUID(manifestation.dataset_id)
-            if dataset_id in devices:
-                # It's already attached here.
-                continue
-            if manifestation.dataset.deleted is True:
-                # Don't attach a dataset to be deleted.
-                continue
-            if manifestation.dataset_id in nonmanifest:
-                volume = _blockdevice_volume_from_datasetid(volumes,
-                                                            dataset_id)
-                if volume is None:
-                    # We know something is non-manifest, but lack any info
```

Comment:
This edge case occurred because we used the data on non-manifest datasets from the control node, but the list of block-device volumes from discovery. Those could get out of sync leading to this edge case. With this change, we only ever look at state discovered locally to make decisions, so the edge case no longer exists.